drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -62,37 +62,74 @@ int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
 			  uint16_t *data)
 {
 	int32_t ret;
-	uint8_t buff[] = {0, 0, 0, 0};
+	uint8_t buff[] = {0, 0};
+
+	ret = adpd410x_reg_read_bytes(dev, address, (uint8_t *) buff, 2);
+	if (ret != SUCCESS) {
+		return ret;
+	}
+
+	*data = ((uint16_t)buff[0] << 8) & 0xff00;
+	*data |= buff[1] & 0xff;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Read a specified number of bytes from device register.
+ * @param dev - Device handler.
+ * @param address - Register address.
+ * @param data - Pointer to the register value container.
+ * @param num_bytes - number of bytes to read
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_reg_read_bytes(struct adpd410x_dev *dev, uint16_t address,
+				uint8_t *data, uint16_t num_bytes)
+{
+	int32_t ret;
+	uint8_t i;
+	uint8_t *buff;
 
 	switch (dev->dev_type) {
 	case ADPD4100:
+		buff = (uint8_t *) calloc(num_bytes + 2, sizeof(*buff));
 		buff[0] = field_get(ADPD410X_UPPDER_BYTE_SPI_MASK, address);
 		buff[1] = (address << 1) & ADPD410X_LOWER_BYTE_SPI_MASK;
 
-		ret = spi_write_and_read(dev->dev_ops.spi_phy_dev, buff, 4);
-		if(ret != SUCCESS)
+		ret = spi_write_and_read(dev->dev_ops.spi_phy_dev, buff, num_bytes + 2);
+		if(ret != SUCCESS) {
+			free(buff);
 			return ret;
+		}
+		for (i = 0; i < num_bytes; i++) {
+			data[i] = buff[i + 2];
+		}
 		break;
 	case ADPD4101:
+		// Number of bytes for an I2C read is an 8-bit number, or at most 255
+		num_bytes = min(num_bytes, 255);
```

Comment:
I won't do this. Return failure if num_bytes > 255 or read all bytes inside a for.
A user will expect all values to be read.